pennylane/templates/decorator.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the template decorator.
+"""
+from pennylane.utils import OperationRecorder
+
+def template(func):
```

Comment:
This solution seems really neat!!

I'm wondering about loaded templates, those that users would be loading from another framework. What could they use if they wanted to invert such templates, specify a decorated function there? (This might be out of the scope of this PR, just came to my mind)

Reply:
That's a good point. I haven't thought of that yet.  

If the loading only returns a function one can just instead return `template(func)` and have the same functionality. But if the statement returns an object we have the object to return the list of applied operations on being called. But this is true no matter which implementation we settle for here.